### PR TITLE
feat: support for GET method for existing servers

### LINKED ISSUE #121: Support GET requests
Right now apolloServer supports only POST requests, which makes its code really simple. However, sometimes it can be useful to send a GET request to a GraphQL server, because more devices support them, and because they make HTTP caching mechanisms easy to use. We should also support those use-cases.

Arguments should either all be in the URL, or all in the body. It shouldn't be possible to mix the two. Maybe we should say that if it's a POST request it has to be in the body, and if it's a GET request, it has to be in the URL?

### PR DESCRIPTION
**Pull request by DxCx**
TODO:
- [X] Support for Express/Connect
- [x] Support for HAPI
- [x] Support for KOA
- [x] Fix Coverage (waiting implementation approval)
- [x] More tests? (need ideas :))
- [ ] Documentation updates? (@helfer?)
- [x] Update CHANGELOG.md with your change (include reference to issue & this PR)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass

Closes #121 , #186

### CODE REVIEW

**Comment by DxCx** [comment]
@helfer @stubailo any update on this?

**Comment by helfer** [comment]
@DxCx Thanks, I'll review it before the weekend!

File: packages/graphql-server-express/src/expressApollo.ts [hunk 1]
```diff
@@ -42,33 +42,48 @@ export function graphqlExpress(options: GraphQLOptions | ExpressGraphQLOptionsFu
     }
 
     const formatErrorFn = optionsObject.formatError || graphql.formatError;
+    let buffer;
```
Comment (helfer) [inline packages/graphql-server-express/src/expressApollo.ts]: can we call this `requestPayload` or something else that's more descriptive than `buffer`?

File: packages/graphql-server-hapi/src/hapiApollo.ts [hunk 2]
```diff
@@ -71,23 +77,44 @@ graphqlHapi.attributes = {
...
+}
 
+function assignIsBatch(method, buffer, reply) {
   // TODO: do something different here if the body is an array.
   // Throw an error if body isn't either array or object.
-  reply(payload && Array.isArray(payload));
+
+  switch ( method ) {
+      case 'get':
+          return reply(false);
```
Comment (helfer) [inline packages/graphql-server-hapi/src/hapiApollo.ts]: We should document the fact that you cannot send batched queries with a GET request.

File: packages/graphql-server-hapi/src/hapiApollo.ts [hunk 3]
```diff
@@ -71,23 +77,44 @@ graphqlHapi.attributes = {
...
+      case 'post':
+          return reply(Array.isArray(buffer));
+      default:
+          throw new Error(`Invalid case reached, method is ${method}`);
+  }
 }
 
-function getGraphQLParams(payload, isBatch, reply) {
+function getGraphQLParams(buffer, isBatch, reply) {
```
Comment (helfer) [inline packages/graphql-server-hapi/src/hapiApollo.ts]: Why rename this variable?

File: packages/graphql-server-express/src/expressApollo.ts [hunk 4]
```diff
@@ -42,33 +42,48 @@ export function graphqlExpress(options: GraphQLOptions | ExpressGraphQLOptionsFu
...
-      res.statusCode = 405;
-      res.write('Apollo Server supports only POST requests.');
-      res.end();
-      return;
-    }
+        buffer = req.body;
+        break;
+      case 'GET':
+        if ( !req.query ) {
+          res.statusCode = 500;
```
Comment (helfer) [inline packages/graphql-server-express/src/expressApollo.ts]: Shouldn't that be a 400?

File: packages/graphql-server-hapi/src/hapiApollo.ts (hunk 2 above)
Comment (DxCx) [inline packages/graphql-server-hapi/src/hapiApollo.ts]: Ack, Where?
i just think that opening batching for GET is going to get monstrous URLs..

**Comment by DxCx** [comment]
hi @helfer just seen your comments, i've replied on some and +1 on some.
please answer and i'll fix and rebase :)

**Comment by DxCx** [comment]
@helfer / @stubailo how do we progress?
is the implementation looks ok?
there are still unresolved review notes..

as far as i can see we need to extend testings to coverage all new logics
and we are good to go.

File: packages/graphql-server-hapi/src/hapiApollo.ts (hunk 2 above)
Comment (helfer) [inline packages/graphql-server-hapi/src/hapiApollo.ts]: Yeah, but I don't think monstrous URLs should be our problem. There's a real risk for that already with normal queries. Most servers don't support URLs longer than 2000 characters, and I think even the introspection query might be longer than that. But that's really not something we should worry about. Maybe we can just put a warning in the README.

File: packages/graphql-server-express/src/expressApollo.ts [hunk 5]
```diff
@@ -42,33 +42,48 @@ export function graphqlExpress(options: GraphQLOptions | ExpressGraphQLOptionsFu
...
-      res.statusCode = 405;
-      res.write('Apollo Server supports only POST requests.');
-      res.end();
-      return;
-    }
+        requestPayload = req.body;
+        break;
+      case 'GET':
+        if ( !req.query ) {
+          res.statusCode = 500;
```
Comment (helfer) [inline packages/graphql-server-express/src/expressApollo.ts]: Yeah, I think 400 is better, because in the case of GET, it means a request was sent without a query, which is definitely not an internal server error. Although... that reminds me that we should also think about how prepared queries would work with GET. I definitely think they should be supported as well, so either this test is in the wrong place, or prepared queries should be a middleware that adds the query to the request before it hits graphql-server.

**Comment by helfer** [comment]
@DxCx looks 99% good to me. Let's add tests and then merge this. The PR has certainly been waiting for long enough!

File: packages/graphql-server-express/src/expressApollo.ts (hunk 5 above)
Comment (DxCx) [inline packages/graphql-server-express/src/expressApollo.ts]: ohhh
i think you are abit confuse :)

if you will take a deeper look, req.query for GET is equal to req.body of POST.
which means, in req.query you will get all the query params as key value.
(?key=value&seckey=value) will become:
```
{
  "key": "value",
  "seckey": "value",
}
```

if i was checking req.query.query, you were correct on your assumption.
is it more clear now? do you still think it should be 400 and not 500?

Comment (DxCx) [inline packages/graphql-server-express/src/expressApollo.ts]: @helfer ?

Comment (helfer) [inline packages/graphql-server-express/src/expressApollo.ts]: @DxCx right, but you're checking req.query. If that's not present, it means it wasn't supplied by the user. In that case I would expect a 400, not a 500.

**Comment by helfer** [comment]
@DxCx can you check off all the boxes that can be checked off, and then tell me what else needs to be done before we can merge this? I think GET support would be a nice present for the end of the year 😄

**Comment by helfer** [comment]
@DxCx okay, sounds good. I think someone else could definitely help or even finish it completely, but for that you should first write down a more detailed list of what still needs to be done. I don't think it would take too long to write that down. 🙂

**Comment by DxCx** [comment]
just make sure all new code is covered by spec file,
and maybe add some more edge cases to the test suite case that is relevant to get (just need to think of some)

**Comment by DxCx** [comment]
hi @helfer we are ready on my side.

File: packages/graphql-server-hapi/src/hapiApollo.ts (hunk 2 above)
Comment (helfer) [inline packages/graphql-server-hapi/src/hapiApollo.ts]: Is there a good reason (from the server's perspective) not to allow batched requests via GET?

**Comment by DxCx** [comment]
cool.

Yeah i think we spoke about it before (either here or on slack)
but, giving the ability to create batched requests over query will create monstrous urls.
i think that at least for start we should block it,
then we can experiment more with it and understand how to allow this in the best way.

**Comment by DxCx** [comment]
i did not understand why not allowing mutation via GET?

**Comment by helfer** [comment]
@DxCx The reason express-graphql doesn't allow mutations over GET is that GET requests in HTTP aren't supposed to have any side-effects. Mutations have side-effects by definition, so express-graphql doesn't allow mutations via GET.

**Comment by DxCx** [comment]
Ok @helfer, sounds right. so you want me to block mutation for all GET requests?

**Comment by DxCx** [comment]
@helfer?

**Comment by helfer** [comment]
@DxCx sorry I dropped the ball on this one. Yes, I think blocking mutations on GET would be good, just to match the behavior of express-graphql. Once you're done with that, please ping me on slack and I will take the PR from there and make sure it gets merged soon. Thanks a lot!

**Comment by DxCx** [comment]
@helfer, no problem, so, we should return 405 error and header says that only POST is allowed..
i guess we should 'lock' get for query and not explicitly block mutation, do you agree?
if so will take care of it in the next 24 hours ;)

**Comment by DxCx** [comment]
@helfer , please review the last 3 commits from today (the rest is just rebase).
i've moved http logic into core package, then added guard for non-query on GET.

File: packages/graphql-server-core/src/runHttpQuery.ts [hunk 6]
```diff
@@ -68,6 +70,13 @@ export async function runHttpQuery(handlerArguments: Array<any>, request: HttpQu
 
   let responses: Array<ExecutionResult> = [];
   for (let requestParams of requestPayload) {
+    if ( isGetRequest && !requestParams.query.trim().startsWith('query')) {
```
Comment (helfer) [inline packages/graphql-server-core/src/runHttpQuery.ts]: According to the spec the query keyword is optional. I know it's annoying, but that's how it is. So we'll have to check here whether it's an anonymous query that starts with `{`.

Comment (stubailo) [inline packages/graphql-server-core/src/runHttpQuery.ts]: Also, the query isn't necessarily the first thing in the document right? For example:

```
fragment X on Y { ... }

{ ... }
```

Is also a valid query.

Comment (stubailo) [inline packages/graphql-server-core/src/runHttpQuery.ts]: I would check for "document doesn't contain the keyword `mutation` outside of all brackets"

Comment (DxCx) [inline packages/graphql-server-core/src/runHttpQuery.ts]: @helfer => i was already figured it out, and actually, you were reviewing outdated chunk, which this was already fixed.
@stubailo => about the fragment, this is really usecase i didn't take into consideration,
i'll try to use GraphQL Parser instead to get the exact operation GraphQL will try to execute.

**Comment by DxCx** [comment]
ok @stubailo @helfer 
please check latest commit which is using GraphQL parser to determinate the query operation.

**Comment by helfer** [comment]
@DxCx Perfect, I think parsing the query is great. And thanks to how the core works it won't be parsed twice. I'll update the docs and roll a new release!
